### DISCUSSION TITLE
Implementing a Mixed Active and Non-Active Brownian Particle Simulation in Hoomd-blue: Configuration and Resizing

### DISCUSSION THREAD

**Discussion by idop11** [glotzerlab/hoomd-blue#1735] [Support] [answered]
Hello,

I am new to using Hoomd-blue and I am attempting to set up a simulation with the following specifications:
I aim to construct a rectangle with a height of 2L and a width of L. In the upper half of this rectangle, I plan to distribute N=1000 active Brownian particles. These particles are to be positioned within a LxL area, based on a predefined configuration stored in a ```.gsd``` file, where each particle is assigned a specific location. In the lower half of the rectangle, I intend to place 500 non-active Brownian particles, also arranged based on a specific LxL configuration from a different ```.gsd``` file.

My question concerns the implementation of this setup. Should I first introduce the 1000 particles within an LxL box, then expand the box to a 2LxL rectangle before adding the additional 500 particles? If this is the case, which functions would allow me to do so? I have read in the documentation about using ```hoomd.update.BoxResize``` for resizing the simulation box, but I need this resizing to happen instantly. Moreover, how can I manage the addition of particles so that, after placing the first 1000 particles, I can incorporate the subsequent 500 particles from the second configuration?
To clarify, I am not seeking a complete code for this setup, but rather guidance on the methodology and the specific functions that would enable me to achieve the intended configuration.

Thank you!

**Comment by joaander** [comment, marked as answer]
All your particle positions are pre-defined, so there is no need to involve HOOMD-blue when placing them. Use your preferred method to place items in a list when constructing the gsd file. For example:
```
frame.configuration.box = [L, 2*L, 0, 0, 0, 0]
frame.particles.position = active_brownian_particle_positions + non_active_particle_positions
```
Assuming you would like the active and non-active particles to have different types:
```
frame.particles.typeid = [0] * len(active_brownian_particle_positions) + [1] * len(non_active_particle_positions)
```

**Reply by idop11** [reply]
Thank you very much, Joshua.

As a follow-up question regarding this setup (I preferred not to open a new discussion, sorry if I should have), I want to initiate this configuration and then let the particles move freely in space. Specifically, I aim for the scenario where if an active particle from the upper half crosses to the lower half (its value on the y-axis changes from positive to negative), it will be changed to a non-active Brownian particle. Similarly, if a non-active Brownian particle from the lower half crosses to the upper half, it should change to an active particle. I thought about implementing this situation using a CustomUpdater (as mentioned in the snipped section below), but I was wondering if there is a more efficient way to achieve this. The current method is time-consuming, and although I know I can adjust the trigger, I'm interested in exploring alternative approaches.

How I implemented it:

```class TypeSwitchAction(hoomd.custom.Action):
    def act(self, timestep):
        # Get a snapshot of the current state of the simulation
        snap = self._state.get_snapshot()

        if snap.communicator.rank == 0:
            # Loop through all particles in the snapshot
            for i in range(snap.particles.N):
                # Check if the particle is in the upper half
                if snap.particles.position[i, 1] > 0:
                    # Change type to 'active' if in the upper half
                    snap.particles.typeid[i] = 0
                else:
                    # Change type to 'non-active' if in the lower half
                    snap.particles.typeid[i] = 1
            # Update the snapshot to apply changes
            self._state.set_snapshot(snap)

if __name__ == "__main__":
    # Initiation and implementation of the simulation, details omitted for brevity

    TypeSwitchUpdater = hoomd.update.CustomUpdater(action=TypeSwitchAction(),
                                                   trigger=hoomd.trigger.Periodic(1))
    simulation.operations += TypeSwitchUpdater

```

**Reply by joaander** [reply]
Yes, `get_snapshot` / `set_snapshot` are slow. Use a local snapshot for better for performance. If your model allows, increase the trigger period. If you need the trigger period to be 1, consider implementing your method in C++ for the best performance: https://github.com/glotzerlab/hoomd-component-template/

**Reply by idop11** [reply]
Thank you.
Just to verify, did you mean that I should use local snapshot in the way mentioned below? Because it actually made the script slower:

```class TypeSwitchAction(hoomd.custom.Action):
    def act(self, timestep):
        with self._state.cpu_local_snapshot as snap:
            for i in range(8000):
                if snap.particles.position[i, 1] > 0:
                    snap.particles.typeid[i] = 0  # Active
                else:
                    snap.particles.typeid[i] = 1  # Non-active
```

**Reply by joaander** [reply]
I would leverage numpy for the best possible performance in a Python solution:
```
with self._state.cpu_local_snapshot as snap:
    snap.particles.typeid[snap.particles.position[:, 1] > 0] = 0
    snap.particles.typeid[snap.particles.position[:, 1] <= 0] = 1
```

**Reply by idop11** [reply]
Thanks, Joshua.

I followed your instructions, and used a local snapshot approach like this:
```
        with self._state.gpu_local_snapshot as snap:
            snap.particles.typeid[snap.particles.position[:, 1] > 0] = 0
            snap.particles.typeid[snap.particles.position[:, 1] <= 0] = 1
```

Instead of the previous method:
```
        snap = self._state.get_snapshot()

        if snap.communicator.rank == 0:
            snap.particles.typeid[snap.particles.position[:, 1] > 0] = 0
            snap.particles.typeid[snap.particles.position[:, 1] <= 0] = 1

            self._state.set_snapshot(snap)
```

However, this resulted in quite different configurations than expected. In this regard, in my initial message, I mentioned that I placed 500 non-active Brownian particles in the lower half. When adjusting the value of kT for these particles in the lower half, I anticipated different outcomes when executing the simulation in terms of how many particles crossed to the upper half after many steps (and therefore changing their type to active Brownian particles), as larger fluctuations would lead to more particles crossing over. However, when using the local snapshot method, I noticed that the final configurations remained the same regardless of the kT setting.

I apologize if my explanation or the details of my script were insufficient, and I'm more than willing to provide further information if necessary. I'm just curious if this issue could somehow be related to the bug mentioned in this discussion: https://github.com/glotzerlab/hoomd-blue/discussions/1721?

**Reply by joaander** [reply]
I can't think of any obvious reason why those two code snippets would behave differently. To make even a halfway decent guess, I need to know how you use the types of the particles in the simulation. I would prefer a *minimal* and complete code example that demonstrates the problem.

**Reply by idop11** [reply]
Sure, I've attached the code. Please note the difference between the two versions of the methods, marked as Option 1 and Option 2.

In addition, I've included the final configurations obtained using kT values of 0.1 and 1 for each of the two methods. Pay attention to the significant differences between the configurations obtained with Option 1 compared to those with Option 2. Initially, the active particles occupied the upper half and the non-active Brownian particles occupied the lower half.

[sample_code.txt](https://github.com/glotzerlab/hoomd-blue/files/14574130/sample_code.txt)

Final State, kT=0.1, Option 1: 
<img width="320" alt="Final_State_kT=0 1_Option1" src="https://github.com/glotzerlab/hoomd-blue/assets/127539588/52208f3e-f658-41e5-a6b8-54ef92204594">
Final State, kT=1, Option 1: 
<img width="320" alt="Final_State_kT=1_Option1" src="https://github.com/glotzerlab/hoomd-blue/assets/127539588/492106ee-3906-40a9-b3d7-cb942406c4e9">
Final State, kT=0.1, Option 2: 
<img width="320" alt="Final_State_kT=0 1_Option2" src="https://github.com/glotzerlab/hoomd-blue/assets/127539588/5f23a66c-a801-47db-abec-a79ad727bff1">
Final State, kT=1, Option 2: 
<img width="320" alt="Final_State_kT=1_Option2" src="https://github.com/glotzerlab/hoomd-blue/assets/127539588/e05bbe70-06a2-47b8-8732-eddd64ec5a5a">

The initial state for all the kTs and the 2 options (note that the initial configuration of the upper half is determined by a specific configuration, where the lower half is by randomly positioning the Brownian particles at this half):
<img width="320" alt="Initial_State" src="https://github.com/glotzerlab/hoomd-blue/assets/127539588/8f28cf15-473e-4f40-90df-50bf8bf1eac5">

**Reply by joaander** [reply]
I don't see a `FilterUpdater` in your code. Without it, your `type_A` and `type_B` filters will always apply to the same particles that they initially matched.

**Reply by idop11** [reply]
Should I integrate FilterUpdater into my script in this manner:
```
    TypeSwitchUpdater = hoomd.update.CustomUpdater(action=TypeSwitchAction(),
                                                   trigger=hoomd.trigger.Periodic(1))
    simulation.operations += TypeSwitchUpdater

    filters_to_update = [type_A, type_B]

    filter_updater = hoomd.update.FilterUpdater(trigger=hoomd.trigger.Periodic(1), filters=filters_to_update)

    simulation.operations.updaters.append(filter_updater)
```
Where is ```TypeSwitchAction``` is:
```
class TypeSwitchAction(hoomd.custom.Action):
    def act(self, timestep):
        with self._state.cpu_local_snapshot as snap:
            snap.particles.typeid[snap.particles.position[:, 1] > 0] = 0
            snap.particles.typeid[snap.particles.position[:, 1] <= 0] = 1
```

I apologize if this seems repetitive, but I couldn't find any past discussions regarding scripts incorporating ```FilterUpdater```. I just want to ensure that I implement my code correctly.

**Reply by idop11** [reply]
Hi Joshua,
Did you have a chance to look over the way I integrated FilterUpdater? Because it still seems that when a particle crosses to the other side it doesn't change its type (its color stays the same as before it went to the other side).

**Reply by joaander** [reply]
I don't know how you compute the color of the particles, so I can't say for certain. I would guess that you forgot to tell the GSD writer that typeid is dynamic.

**Reply by idop11** [reply]
Indeed, this solved the problem. Thank you.
I have one more question regarding this setup (I hope it's fine that I'm asking multiple questions in this post – I think it's more straightforward than creating a new one).
I want to adapt this system so that if a non-active Brownian particle (from the lower half) crosses to the upper half, then its new self-propelling velocity (the parameter ```vzero```) will align with the direction of the previous velocity vector it had the moment before it crossed into the upper box when it was still non-active. For example, if at the step before changing its type to an active particle, the unit vector of velocity was $(\frac{\sqrt{2}}{2},\frac{\sqrt{2}}{2},0)$, then after crossing, its self-propelling velocity should be: $vzero\cdot(\frac{\sqrt{2}}{2},\frac{\sqrt{2}}{2},0)$.
I understand that it's not possible to change the active force for each particle individually, so I thought about determining the orientation of the Brownian particle before it changed its type, ensuring that after crossing and becoming an active particle, it will retain the same orientation.
However, I haven't managed to figure out how to calculate the orientation of a non-active Brownian particle (that performs a random walk). To clarify, by the orientation of the non-active particle, I mean the direction in which it fluctuates during a specific time step. Is there a straightforward way to find its orientation in this state?

I'm attaching my code once again (I made some minor changes following your advice):
[sample_code_new.txt](https://github.com/glotzerlab/hoomd-blue/files/14704247/sample_code_new.txt)

**Reply by joaander** [reply]
Yes, you should set your particle's orientation to achieve the desired direction of the active vector in the world frame. The active force *vector* you supply to `force.Active` is in the particle's reference frame. You can access the properties `snap.particles.velocity` and `snap.particles.orientation` to implement this. You may find a library like [rowan](https://rowan.readthedocs.io/) useful.

Note that embedding unrelated questions in discussion threads makes it harder for others to discover the question and the answer.

**Reply by idop11** [reply]
Thank you. If I have any new questions, I'll make sure to post them separately.

As for my current question, I've observed that the orientation of non-active Brownian particles remains unchanged from their initial state, despite their orientation ostensibly changing at each timestep (perhaps the fact that their motion is solely due to random thermal fluctuations is influencing this?). Do you have any insights on how to address this issue?

**Reply by joaander** [reply]
Please see the documentation on ["Degrees of freedom"](https://hoomd-blue.readthedocs.io/en/v4.6.0/package-md.html#hoomd.md.Integrator). Your simulation does not meet these conditions:
 * DPDConservative applies 0 torque to particles.
 * You do not set a moment of inertia.
 * You do not set `integrate_rotational_degrees_of_freedom=True`. 

Therefore, the integration method will not integrate the rotational degrees of freedom (orientation and angular_momentum) and they will remain fixed. I don't think this is an issue and it does not need to be addressed. Your particles are spherical!

You previously asked about *setting* a particle's orientation when it switched from non-active to active and I answered that question. Once active, your `ActiveRotationalDiffusion` updater will update the orientations of the active particles to implement rotational diffusion.
